### PR TITLE
read_ncdf proxy

### PR DESCRIPTION
**Pull request by dblodgett-usgs**
This is a WIP for awareness and probably some guidance on how this ought to work. 

My hope is to get proxy behavior in stars for a variety of applications, but my primary one is for iteration of very large datasets where the coordinate variables are the same for many thousand time steps and/or variables. 

This ability to create stars dimension objects from netcdf sources while maintaining a pointer to the file or web service will be very useful.

### CODE REVIEW

**Comment by dblodgett-usgs** [comment]
Making some minor progress... with a print method implemented, I now see:
``` r
library(stars)
#> Loading required package: abind
#> Loading required package: sf
#> Linking to GEOS 3.9.1, GDAL 3.2.1, PROJ 7.2.1
f <- system.file("nc/reduced.nc", package = "stars")
nc <- read_ncdf(f, proxy = TRUE)
#> no 'var' specified, using sst, anom, err, ice
#> other available variables:
#>  lon, lat, zlev, time
#> No projection information found in nc file. 
#>  Coordinate variable units found to be degrees, 
#>  assuming WGS84 Lat/Lon.

nc
#> netcdf source stars_proxy object with 1 attribute in 1 file(s):
#> [[1]]
#> [1] "[...]/reduced.nc"
#> 
#> dimension(s):
#>      from  to offset delta  refsys point         values x/y
#> lon     1 180     -1     2  WGS 84    NA           NULL [x]
#> lat     1  90    -90     2  WGS 84    NA           NULL [y]
#> zlev    1   1     NA    NA      NA    NA              0    
#> time    1   1     NA    NA POSIXct    NA 1981-12-31 UTC    
#> 
#> First nc_request:
#> List of 4
#>  $ start: num [1:4] 1 1 1 1
#>  $ count: num [1:4] 180 90 1 1
#>  $ size : num 16200
#>  $ units: Units: [°C] num 1
```

<sup>Created on 2021-12-21 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by dblodgett-usgs** [comment]
The way I started implementing this is to return a sidecar attribute containing the required netcdf request arguments that could be used to fetch the data later on. 

Starting to think about how this could be used in the existing methods for stars_proxy objects... need to ponder it a bit. With GDAL sourced proxys, the pattern for accessing the proxy is going to be a bit different -- to the point that a different class could even be in order for NetCDF proxy... hopefully not. 

@edzer or anyone else, do you have any thoughts here?

**Comment by edzer** [comment]
> @edzer or anyone else, do you have any thoughts here?

I assume you want different behavior from the current `stars_proxy` (which always goes straight into gdal). Do you want to read sub-arrays using `read_ncdf`? In that case (as a start) it might be easiest to subclass class `c("stars_proxy", "stars")` e.g. by `c("nc_proxy", "stars_proxy", "stars")` - that way you only have to write the methods for `nc_proxy` where they deviate from `stars_proxy`. Does that make sense?

File: R/proxy.R [hunk 1]
```diff
@@ -35,10 +33,51 @@ print.stars_proxy = function(x, ..., n = 1e5, nfiles = 10, simplify = TRUE) {
...
+		g <- 1
+		for(s in 1:length(selecter)) {
+			matched <- sapply(x, function(nc1, nc2) {
+				identical(nc1, nc2)
+			}, nc2 = x[[s]])
+
+			if(any(!matched[1:s])) g <- g + 1
+			selecter[matched] <- g
+		}
+	}
```
Comment (dblodgett-usgs) [inline R/proxy.R:71]: There has GOT to be a better way to find groups of identical objects?

**Comment by dblodgett-usgs** [comment]
Yes it does -- I think that will be a good way to go. I started working that in for a print method that will kind of drive the workflow. With the above, we now have: 
``` r
library(stars)
#> Loading required package: abind
#> Loading required package: sf
#> Linking to GEOS 3.9.1, GDAL 3.2.1, PROJ 7.2.1
f <- system.file("nc/reduced.nc", package = "stars")
nc <- read_ncdf(f, proxy = TRUE)
#> no 'var' specified, using sst, anom, err, ice
#> other available variables:
#>  lon, lat, zlev, time
#> No projection information found in nc file. 
#>  Coordinate variable units found to be degrees, 
#>  assuming WGS84 Lat/Lon.
nc
#> netcdf source stars proxy object from:
#> [1] "[...]/reduced.nc"
#> 
#> Available nc variables:
#> sst  [°C] 
#> anom  [°C] 
#> err  [°C] 
#> ice  [percent] 
#> 
#> dimension(s):
#>      from  to offset delta  refsys point         values x/y
#> lon     1 180     -1     2  WGS 84    NA           NULL [x]
#> lat     1  90    -90     2  WGS 84    NA           NULL [y]
#> zlev    1   1     NA    NA      NA    NA              0    
#> time    1   1     NA    NA POSIXct    NA 1981-12-31 UTC    
#> 
#> 
#> nc_request:
#> 
#> Variables: sst, anom, err
#>  Size: 16200 cells
#>  Units: °C
#>  Start: 1, 1, 1, 1
#>  Count: 180, 90, 1, 1
#>  Dimension Match: 1, 2, 3, 4
#> Variables: ice
#>  Size: 16200 cells
#>  Units: percent
#>  Start: 1, 1, 1, 1
#>  Count: 180, 90, 1, 1
#>  Dimension Match: 1, 2, 3, 4
```

<sup>Created on 2021-12-22 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

I started to summarize how this will be used in a vignette for the sake of getting it written down. Not sure that's the right way to go but it's there anyways.

**Comment by dblodgett-usgs** [comment]
p.s. I have no idea why my git or rstudio is changing whitespace in the files I touch. Will try and clean that up before this is all done.

**Comment by kadyb** [comment]
@dblodgett-usgs, this can be changed in RStudio settings:
_Global_ or _Project options_ > _Code_ > _Strip trailing horizontal whitespace when saving_

But I see this option is enabled in this project. Maybe there should be a separate PR to remove these whitespaces (it will be easier to see changes in the right PR)?

**Comment by edzer** [comment]
> But I see this option is enabled in this project.

Which is wrong: I don't use RStudio when coding, someone else has uploaded the .Rproj file with the wrong setting.

**Comment by dblodgett-usgs** [comment]
I will attempt a re-base.

**Comment by dblodgett-usgs** [comment]
OK -- cleaner now. Sorry for the churn. Could probably close this PR sooner than later in favor of a new cleaned up one.

**Pull request closed without merging**